
---
title: "Log-linear model with categorical variables"
author: "Jim Shapleigh"
edited by: "Sarah Nadeau"
date: "September 1, 2016"
updated: "February 23, 2018"
output: html_document
---

Set working directory to "Data"" directory
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/nadeau/Box\ Sync/AK_research/Data")
```

```{r, include=FALSE}
library(plyr)
library(ggplot2)
library(readr)
library(dplyr)
library(magrittr)
library(tidyr)
library(tibble)
library(viridis)
library(knitr)
library(reshape2)
library(emmeans)
library(multcompView)
library(lme4)
library(forcats)
library(grid)
library(gridExtra)
```
Load data
NOTE: read counts generated from FMAP, genes are reported using KEGG orthologies

```{r}

fmap.df = read_csv("dn_read_abundance_csv.csv")
print(fmap.df)
attach(fmap.df)

```
Check data distribution
-it does appear poisson in distribution

```{r}
hist(fmap.df$Nar)
hist(fmap.df$Nap)
hist(fmap.df$NirS)
hist(fmap.df$NirK)
hist(fmap.df$NorBC)
hist(fmap.df$NosZ)
```

Define GLM model fit

```{r}

fit = function(x) {
  fit = glm(x ~ site_code + offset(log(read_depth/10^6)) - 1, quasipoisson)
}

n_sites = 6
n_genes = 6
```
Initialize data frame called 'to_plot' for figure

```{r}
to_plot = data.frame(matrix(nrow = 6*6, ncol = 8))
colnames(to_plot) <- c("gene", "site", "mean", "ci_low", "ci_high", "new","ord", "group")
to_plot$site = c('U','U','U','U','U','U',
                 'PFO','PFO','PFO','PFO','PFO','PFO',
                 'PSS','PSS','PSS','PSS','PSS','PSS',
                 'PEM','PEM','PEM','PEM','PEM','PEM', 
                 'RNF','RNF','RNF','RNF','RNF','RNF',
                 'RF','RF','RF','RF','RF','RF')

to_plot$gene = c('Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ',
                 'Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ',
                 'Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ',
                 'Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ',
                 'Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ',
                 'Nar', 'Nap', 'NirS', 'NirK', 'NorBC', 'NosZ')
to_plot$ord = to_plot$gene
to_plot$new = to_plot$site
```
Fit model for all genes, record predicted counts in fmap.df

```{r}

Nar_model = fit(Nar)
Nar_summary = summary.glm(Nar_model)
fmap.df$Nar_predicted <- exp(predict.glm(Nar_model, fmap.df))

Nap_model = fit(Nap)
Nap_summary = summary.glm(Nap_model)
fmap.df$Nap_predicted <- exp(predict.glm(Nap_model, fmap.df))

NirS_model = fit(NirS)
NirS_summary = summary.glm(NirS_model)
fmap.df$NirS_predicted <- exp(predict.glm(NirS_model, fmap.df))

NirK_model = fit(NirK)
NirK_summary = summary.glm(NirK_model)
fmap.df$NirK_predicted <- exp(predict.glm(NirK_model, fmap.df))

NorBC_model = fit(NorBC)
NorBC_summary = summary.glm(NorBC_model)
fmap.df$NorBC_predicted <- exp(predict.glm(NorBC_model, fmap.df))

NosZ_model = fit(NosZ)
NosZ_summary = summary.glm(NosZ_model)
fmap.df$NosZ_predicted <- exp(predict.glm(NosZ_model, fmap.df))

to_plot_list <- split(fmap.df, as.factor(fmap.df$site_code))
```
Calculate model mean and confidence interval bounds for each gene at each site
Mean is Estimated Marginal Mean

```{r}
gene_name = 'Nap'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(Nap_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)

# check model fit and assumptions
plot(fmap.df$Nap, exp(predict(Nap_model)))
abline(a = 0, b = 1)
plot(resid(Nap_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}
```
``` {r}
gene_name = 'Nar'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(Nar_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)

# check model fit and assumptions
plot(fmap.df$Nar, exp(predict(Nar_model)))
abline(a = 0, b = 1)
plot(resid(Nar_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}

```
NOTE: nirS model fits very poorly for some sites

UCL calculation for PEM:
confidence interval = critical value * std error (this formula holds for large values of n)
t-value for df=Inf and alpha = 0.05 (two-tailed) is 1.960 (critical value)

UCL = critical value * std error 
    = 1.96 x 4321.2749 
    = 8469
exp(8469) = Inf

Standard error is too high to report the model prediction, so not included in plot.
Should think about whether to include as a * at zero or not
``` {r}
gene_name = 'NirS'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(NirS_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)
print(summary(NirS_model))

# check model fit and assumptions
plot(fmap.df$NirS, exp(predict(NirS_model)))
abline(a = 0, b = 1)
plot(resid(NirS_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  
  if (emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] != Inf) {
    to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
    
    to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
    
    to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  }
  
  # else {
  #   to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = 0
  #   
  #   to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = 0 
  #   
  #   to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = 0
  # }
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}
```
``` {r}
gene_name = 'NirK'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(NirK_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)

# check model fit and assumptions
plot(fmap.df$NirK, exp(predict(NirK_model)))
abline(a = 0, b = 1)
plot(resid(NirK_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}
```
``` {r}
gene_name = 'NorBC'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(NorBC_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)

# check model fit and assumptions
plot(fmap.df$NorBC, exp(predict(NorBC_model)))
abline(a = 0, b = 1)
plot(resid(NorBC_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}
```
``` {r}
gene_name = 'NosZ'

# report margnial means and confidence interval for model predicted gene counts
emm=emmeans(NosZ_model, "site_code", type="response")
emm.summary = summary(emm)
with_cld = cld(emm)
print(with_cld)

# check model fit and assumptions
plot(fmap.df$NosZ, exp(predict(NosZ_model)))
abline(a = 0, b = 1)
plot(resid(NosZ_model))
abline(a = 0, b = 0)

# add mean predicted gene count, confidence limits to to_plot data frame
for (site in c("U", "PFO", "PSS", "PEM", "RNF", "RF")) {
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'mean'] = emm.summary[ which(emm.summary$site_code == site), 'rate']
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_high'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.UCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'ci_low'] = emm.summary[ which(emm.summary$site_code == site), 'asymp.LCL'] 
  
  to_plot[ which(to_plot$gene == gene_name  & to_plot$site == site), 'group'] = with_cld[ which(with_cld$site_code == site), '.group']
}
```
Make plot
NOTE: NirS at sites PSS, PEM, U not estimated from model because standard error too high

```{r}

# to_plot$ord = to_plot$gene
# to_plot$new = to_plot$site
# 
# p = ggplot(to_plot, aes(mean, new, colour=gene)) + geom_point(size=.5)
# 
# p = p +  geom_errorbarh(aes(xmax = ci_high, xmin = ci_low, height = .75))
# 
# p = p + facet_grid(ord ~ .)
# 
# p = p + theme(panel.background = element_rect(fill = "white"), panel.grid.major.x = element_blank(), panel.grid.major.y = element_line(linetype=3, color="light grey"), legend.position="none") #add legend position none.
# 
# p = p + labs(x = "Normalized read count x10^6 \n Bars indicate 95% confidence interval. \n Numbers indicate significantly different groupings, p < 0.05. \n", y = "Site", size=8)
# 
# p = p +  geom_text(aes(ci_high, new, label=group, hjust = -0.15), size=2.5)
#  
# p = p + coord_fixed(ratio = 4)
# 
# p = p + theme(text = element_text(size = 8))
# 
# p
# 
# ggsave("plot.png", width=5, height=5)
# 
# # tiff("facet reads.tiff", width = 4, height = 5.5, units = 'in', res = 300)
# plot(p) # Make plot
# dev.off()
```
Plot highlighting Riparian with Fish (RF) sites

```{r}

to_plot$ord = to_plot$gene
to_plot$new = to_plot$site

to_plot <- to_plot %>%
  group_by(site) 

to_plot_with_fish <- to_plot %>%
  filter(site == 'RF')

p = ggplot(to_plot, aes(x = mean, y = new, colour=gene)) + 
  geom_point(size = .5, alpha = .4) +
  geom_errorbarh(aes(xmax = ci_high, xmin = ci_low, height = .75), alpha = .4) +
  geom_point(data = to_plot_with_fish, size = .5, alpha = 1) +
  geom_errorbarh(data = to_plot_with_fish, aes(xmax = ci_high, xmin = ci_low, height = .75, alpha = 1))

p = p + facet_grid(ord ~ .)

p = p + theme(panel.background = element_rect(fill = "white"), panel.grid.major.x = element_blank(), panel.grid.major.y = element_line(linetype=3, color="light grey"), legend.position="none") #add legend position none.

p = p + labs(x = "Reads per Million Total Reads \n bars indicate 95% confidence interval. \n numbers indicate significantly different groupings, p < 0.05 \n", y = "Site", title = "Riparian sites with salmon characterized by higher relative abundance of denitrification genes.", size=6)

# p = p + grid.arrange(p, right = textGrob("Gene", rot = -90, vjust = 1))
# p = p + arrange(p, right = textGrob("Gene", rot = -90, vjust = 1))

p = p +  geom_text(aes(ci_high, new, label=group, hjust = -0.15), size=2.5, alpha = .4)
p = p +  geom_text(data = to_plot_with_fish, aes(ci_high, new, label=group, hjust = -0.15), size=2.5, alpha = 1)

p = p + coord_fixed(ratio = 4)

p = p + theme(text = element_text(size = 8))

p

# ggsave("plot.png", width=5, height=5)

# tiff("facet reads.tiff", width = 4, height = 5.5, units = 'in', res = 300)
plot(p) # Make plot
dev.off()
```